hourglass_blocks.py

```diff
@@ -0,0 +1,193 @@
+
+    f1 = bottleneck(bottom, num_channels, hgname + '_l1')
+    _x = MaxPool2D(pool_size=(2, 2), strides=(2, 2))(f1)
+
+    f2 = bottleneck(_x, num_channels, hgname + '_l2')
+    _x = MaxPool2D(pool_size=(2, 2), strides=(2, 2))(f2)
+
+    f4 = bottleneck(_x, num_channels, hgname + '_l4')
+    _x = MaxPool2D(pool_size=(2, 2), strides=(2, 2))(f4)
+
+    f8 = bottleneck(_x, num_channels, hgname + '_l8')
+
+    return (f1, f2, f4, f8)
+
+
+def connect_left_to_right(left, right, bottleneck, name, num_channels):
+    '''
+    :param left: connect left feature to right feature
+    :param name: layer name
+    :return:
+    '''
+    # left -> 1 bottlenect
+    # right -> upsampling
+    # Add   -> left + right
+
+    _xleft = bottleneck(left, num_channels, name + '_connect')
+    _xright = UpSampling2D()(right)
+    add = Add()([_xleft, _xright])
+    out = bottleneck(add, num_channels, name + '_connect_conv')
+    return out
+
+
+def bottom_layer(lf8, bottleneck, hgid, num_channels):
+    # blocks in lowest resolution
+    # 3 bottlenect blocks + Add
+
+    lf8_connect = bottleneck(lf8, num_channels, str(hgid) + "_lf8")
+
+    _x = bottleneck(lf8, num_channels, str(hgid) + "_lf8_x1")
+    _x = bottleneck(_x, num_channels, str(hgid) + "_lf8_x2")
+    _x = bottleneck(_x, num_channels, str(hgid) + "_lf8_x3")
+
+    rf8 = Add()([_x, lf8_connect])
+
+    return rf8
+
+
+def create_right_half_blocks(leftfeatures, bottleneck, hglayer, num_channels):
+    lf1, lf2, lf4, lf8 = leftfeatures
+
+    rf8 = bottom_layer(lf8, bottleneck, hglayer, num_channels)
+
+    rf4 = connect_left_to_right(lf4, rf8, bottleneck, 'hg' + str(hglayer) + '_rf4', num_channels)
+
+    rf2 = connect_left_to_right(lf2, rf4, bottleneck, 'hg' + str(hglayer) + '_rf2', num_channels)
+
+    rf1 = connect_left_to_right(lf1, rf2, bottleneck, 'hg' + str(hglayer) + '_rf1', num_channels)
+
+    return rf1
+
+
+def create_heads(prelayerfeatures, rf1, num_classes, hgid, num_channels):
+    # two head, one head to next stage, one head to intermediate features
+    head = Conv2D(num_channels, kernel_size=(1, 1), activation='relu', padding='same', name=str(hgid) + '_conv_1x1_x1')(
+        rf1)
+    head = BatchNormalization()(head)
+
+    # for head as intermediate supervision, use 'linear' as activation.
+    head_parts = Conv2D(num_classes, kernel_size=(1, 1), activation='linear', padding='same',
+                        name=str(hgid) + '_conv_1x1_parts')(head)
+
+    # use linear activation
+    head = Conv2D(num_channels, kernel_size=(1, 1), activation='linear', padding='same',
+                  name=str(hgid) + '_conv_1x1_x2')(head)
+    head_m = Conv2D(num_channels, kernel_size=(1, 1), activation='linear', padding='same',
+                    name=str(hgid) + '_conv_1x1_x3')(head_parts)
+
+    head_next_stage = Add()([head, head_m, prelayerfeatures])
+    return head_next_stage, head_parts
+
+
+def euclidean_loss(x, y):
+    return K.sqrt(K.sum(K.square(x - y)))
```

Comment:
Is this a temporary loss function, will we be replacing this?

Reply:
Temporary for now